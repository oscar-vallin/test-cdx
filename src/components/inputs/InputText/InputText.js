import PropTypes from 'prop-types';
import { StyledTextField } from './InputText.styles';

const InputText = ({
  id = 'InputText',
  type = 'text',
  disabled = false,
  onChange,
  autofocus = true,
  errorMessage,
  onKeyDown,
  onKeyEnter,
  value,
  ...props
}) => {
  const handleKey = (key) => {
    if (key === 'Enter' && onKeyEnter) return onKeyEnter();

    if (onKeyDown) return onKeyDown(key);

    return null;
  };

  return (
    <StyledTextField
      id={id}
      type={type}
      autofocus={autofocus}
      disabled={disabled}
      onChange={onChange}
      onKeyDown={({ key }) => handleKey(key)}
      value={value}
      errorMessage={errorMessage}
      {...props}
    />
  );
};

InputText.propTypes = {
  id: PropTypes.string,
  type: PropTypes.string,
  disabled: PropTypes.bool,
  autofocus: PropTypes.bool,
  onChange: PropTypes.func,
  errorMessage: PropTypes.string,
  value: PropTypes.string,
};

export { InputText };
