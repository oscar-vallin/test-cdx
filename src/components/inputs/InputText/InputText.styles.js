import styled from 'styled-components';
import { TextField } from '@fluentui/react';

export const StyledTextField = styled(TextField)`
  width: 100%;
`;
