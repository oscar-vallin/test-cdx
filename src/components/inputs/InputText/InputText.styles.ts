import styled from 'styled-components';
import { TextField } from '@fluentui/react';

export const StyledTextField = styled(TextField)<StyledTextFieldProps>`
  width: 100%;
`;

type StyledTextFieldProps = {
  autofocus: boolean;
  value: string;
};
