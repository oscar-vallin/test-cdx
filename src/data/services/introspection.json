{
  "possibleTypes": {
    "NVP": [
      "NVPStr",
      "NVPId"
    ],
    "PasswordRule": [
      "PasswordLengthRule",
      "PasswordWhitespaceRule",
      "PasswordCharacterRule",
      "PasswordStrengthRule",
      "PasswordRuleGroup"
    ]
  }
}