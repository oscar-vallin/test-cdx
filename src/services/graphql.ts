import { gql } from '@apollo/client';
import * as Apollo from '@apollo/client';
