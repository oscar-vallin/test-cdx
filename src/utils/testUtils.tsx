import { shallow, mount, render } from 'enzyme';
import { ThemeProvider } from 'styled-components';
import { defaultTheme } from '../styles/themes/index';
import { theme } from '../styles/themes/theme';

// eslint-disable-next-line react/prop-types
const LightTheme = { ...theme, ...defaultTheme };
export const mountWithTheme = (children: any) => mount(<ThemeProvider theme={LightTheme}>{children}</ThemeProvider>);

export const shallowWithTheme = (children: any) =>
  shallow(<ThemeProvider theme={LightTheme}>{children}</ThemeProvider>);

export const renderWithTheme = (children: any) => render(<ThemeProvider theme={LightTheme}>{children}</ThemeProvider>);

export * from '@testing-library/react';

export const randomString = (length: number) => {
  let result = '';
  const characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz';
  const charactersLength = characters.length;
  for (let i = 0; i < length; i++) {
    result += characters.charAt(Math.floor(Math.random() * charactersLength));
  }
  return result;
};
