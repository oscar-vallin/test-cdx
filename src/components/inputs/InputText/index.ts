import { InputText, UIInputText } from './InputText';

export { InputText, UIInputText };
