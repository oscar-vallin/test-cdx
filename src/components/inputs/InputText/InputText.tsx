import React, { ReactElement } from 'react';
import FormLabel, { UIFormLabel } from 'src/components/labels/FormLabel';
import { UiField, UiStringField } from 'src/data/services/graphql';
import { EmptyValue, FieldValue, StyledTextField } from './InputText.styles';

export type InputTextProps = {
  id?: string;
  type?: string;
  disabled?: boolean;
  onChange?: (e: React.FormEvent<HTMLInputElement | HTMLTextAreaElement>, newValue?: string) => void | null;
  autofocus?: boolean;
  errorMessage?: string;
  onKeyDown?: any | null;
  onKeyEnter?: any | null;
  value?: string;
  placeholder?: string;
  renderLabel?: boolean;
  required?: boolean;
  canRevealPassword?: boolean;
  label?: string;
  maxLength?: number;
  minLength?: number;
  info?: string;
  autocomplete?: string
};

const InputText = ({
  id,
  type,
  disabled,
  onChange,
  autofocus,
  errorMessage,
  onKeyDown,
  onKeyEnter,
  value,
  info,
  required,
  renderLabel = true,
  autocomplete,
  ...props
}: InputTextProps): ReactElement => {
  const handleKey = (key) => {
    if (key === 'Enter' && onKeyEnter) return onKeyEnter();

    if (onKeyDown) return onKeyDown(key);

    return null;
  };

  const onRenderLabel = () => {
    if (renderLabel) {
      return <FormLabel id={`${id}_lbl`} required={required} info={info} errorMessage={errorMessage} {...props} />;
    }
    return null;
  };

  return (
    <StyledTextField
      id={id}
      type={type}
      autofocus={autofocus ?? false}
      disabled={disabled}
      onChange={onChange}
      onKeyDown={({ key }) => handleKey(key)}
      value={value ?? ''}
      errorMessage={errorMessage}
      onRenderLabel={onRenderLabel}
      autoComplete={autocomplete}
      {...props}
    />
  );
};

type UIReadOnlyType = {
  id: string;
  uiField?: UiField;
  renderLabel?: boolean;
};

const UIInputTextReadOnly = ({ id, uiField, renderLabel = true }: UIReadOnlyType) => {
  const getValue = () => {
    let text = '';
    if (uiField?.hasOwnProperty('description')) {
      text = uiField['description'];
    } else if (uiField?.hasOwnProperty('value')) {
      text = uiField['value'];
    }
    if (text && text.length > 0) {
      return text;
    }
    return <EmptyValue>&lt;empty&gt;</EmptyValue>;
  };

  return (
    <>
      {renderLabel && <UIFormLabel id={`${id}_lbl`} uiField={uiField} />}
      <FieldValue id={id}>{getValue()}</FieldValue>
    </>
  );
};

type UIInputTextType = {
  id: string;
  uiField?: UiStringField;
  onChange?: (e: React.FormEvent<HTMLInputElement | HTMLTextAreaElement>, newValue?: string) => void | null;
  value?: string;
  placeholder?: string;
  renderLabel?: boolean;
  autocomplete?: string;
};

const UIInputText = ({ id, uiField, onChange, value, placeholder, renderLabel = true, autocomplete}: UIInputTextType) => {
  if (uiField?.readOnly === true) {
    return <UIInputTextReadOnly id={id} uiField={uiField} renderLabel={renderLabel} />;
  }
  return (
    <InputText
      id={id}
      type="text"
      value={value}
      disabled={uiField?.readOnly ?? false}
      autofocus={false}
      label={uiField?.label}
      errorMessage={uiField?.errMsg ?? undefined}
      info={uiField?.info ?? undefined}
      required={uiField?.required ?? false}
      renderLabel={renderLabel}
      onChange={onChange}
      placeholder={placeholder}
      minLength={uiField?.min}
      maxLength={uiField?.max}
      autocomplete={autocomplete}
    />
  );
};

export { InputText, UIInputText, UIInputTextReadOnly };
