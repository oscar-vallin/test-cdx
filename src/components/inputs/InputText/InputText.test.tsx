import React from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import toJSON from 'enzyme-to-json';
import { shallow } from 'enzyme';
import { InputText, UIInputText } from './index';
import { mountWithTheme } from 'src/utils/testUtils';
import { UiStringField } from 'src/data/services/graphql';
import { UIInputTextReadOnly } from './InputText';

const placeholderText = 'This is a placeholder';

const defaultProps = {
  id: 'InputText',
  type: 'text',
  disabled: false,
  onChange: undefined,
  autofocus: true,
  errorMessage: undefined,
  onKeyDown: null,
  onKeyEnter: null,
  value: '',
};

const defaultTestProps = {
  ...defaultProps,
  info: jest.fn(),
};

const nickName: UiStringField = {
  label: 'Nick Name',
  value: '',
  required: true,
  visible: true,
  info: "A Name that isn't your name",

  errMsg: 'This field is Required',
  min: 0,
  max: 60,
};

test('Matches Snapshot', () => {
  const wrapper = shallow(<InputText {...defaultProps} placeholder={placeholderText} />);

  expect(toJSON(wrapper)).toMatchSnapshot();
});

describe('Basic Input Component', () => {
  it('Should renders InputText Component', () => {
    const wrapper = mountWithTheme(<InputText {...defaultProps} />);
    expect(wrapper.find('input')).toHaveLength(1);
  });

  it('Should renders InputText with placeholder', () => {
    const wrapper = shallow(<InputText {...defaultProps} placeholder={placeholderText} />);
    expect(wrapper.prop('placeholder')).toEqual(placeholderText);
  });

  it('Should renders InputText with Props', () => {
    const wrapper = shallow(<InputText {...defaultProps} placeholder={placeholderText} />);
    expect(wrapper.prop('id')).toEqual(defaultProps.id);
    expect(wrapper.prop('type')).toEqual(defaultProps.type);
    expect(wrapper.prop('errorMessage')).toEqual(defaultProps.errorMessage);
    expect(wrapper.prop('value')).toEqual(defaultProps.value);
    expect(wrapper.prop('disabled')).not.toBeTruthy();
    expect(wrapper.prop('autofocus')).toBeTruthy();
    expect(wrapper.prop('placeholder')).toEqual(placeholderText);
  });

  it('@Testing: Render Input', () => {
    const wrapper = mountWithTheme(<InputText {...defaultProps} placeholder={placeholderText} />);
    expect(wrapper.html().indexOf(placeholderText)).toBeGreaterThan(-1);
  });

  it('@Testing: Render by Role', () => {
    const wrapper = mountWithTheme(<InputText {...defaultProps} placeholder={placeholderText} />);
    expect(wrapper.html().indexOf('type="text"')).toBeGreaterThan(-1);
  });

  it('@Testing: Input Text', () => {
    let updatedValue = '';
    const onChange = (e, newValue) => {
      updatedValue = newValue;
    };

    const wrapper = mountWithTheme(<InputText {...defaultProps} onChange={onChange} />);
    expect(wrapper.find('input')).toHaveLength(1);
    wrapper.find('input').simulate('change', { target: { value: 'searchString' } });
    expect(updatedValue).toEqual('searchString');
  });

  it('@Testing: Check call function when key press = Enter', () => {
    const mockFn = jest.fn();
    const mockFn2 = jest.fn();
    const wrapper = mountWithTheme(
      <InputText
        {...defaultProps}
        placeholder={placeholderText}
        onKeyDown={(key) => mockFn(key)}
        onKeyEnter={mockFn2}
      />
    );
    wrapper.find('input').simulate('keyDown', { key: 'Enter' });

    expect(mockFn2).toHaveBeenCalled();
  });

  it('@Testing: Check call function when key press', () => {
    const mockFn = jest.fn();
    const mockFn2 = jest.fn();
    const wrapper = mountWithTheme(
      <InputText
        {...defaultProps}
        placeholder={placeholderText}
        onKeyDown={(key) => mockFn(key)}
        onKeyEnter={mockFn2}
      />
    );
    wrapper.find('input').simulate('keyDown', { key: 'K' });

    expect(mockFn).toHaveBeenCalled();
  });

  it('@Testing: Check call function when key press', () => {
    const mockFn = jest.fn();

    const wrapper = mountWithTheme(<InputText {...defaultProps} placeholder={placeholderText} />);

    wrapper.find('input').simulate('keyDown');

    expect(mockFn).not.toHaveBeenCalled();
  });

  it('Should send the property info to the component', () => {
    const tree = mountWithTheme(<InputText {...defaultTestProps} info="Tooltip info" />);
    expect(tree).toMatchSnapshot();
  });

  it('UIInput Text Rendering', () => {
    const wrapper = shallow(<UIInputText id="nickname" uiField={nickName} value="Jimbo" />);

    expect(wrapper.find('InputText')).toHaveLength(1);
    expect(wrapper.find('InputText').get(0).props.autofocus).toEqual(false);
    expect(wrapper.find('InputText').get(0).props.value).toEqual('Jimbo');
    expect(wrapper.find('InputText').get(0).props.type).toEqual('text');
    expect(wrapper.find('InputText').get(0).props.info).toEqual("A Name that isn't your name");
    expect(wrapper.find('InputText').get(0).props.required).toEqual(true);
    expect(wrapper.find('InputText').get(0).props.errorMessage).toEqual('This field is Required');
    expect(wrapper.find('InputText').get(0).props.minLength).toEqual(0);
    expect(wrapper.find('InputText').get(0).props.maxLength).toEqual(60);
  });

  it('Test OnChange', () => {
    let update: string = '';
    const onChange = (e: React.FormEvent<HTMLInputElement | HTMLTextAreaElement>, newValue?: string) => {
      update = newValue ?? '';
    };

    const wrapper = mountWithTheme(<UIInputText id="foo" uiField={nickName} value="RZA" onChange={onChange} />);

    expect(wrapper.find('input')).toHaveLength(1);
    wrapper.find('input').simulate('change', { target: { value: 'Skylar' } });

    expect(update).toEqual('Skylar');
  });

  it('Test Read Only Field', () => {
    const roNickName: UiStringField = {
      ...nickName,
      value: 'Kerman',
      readOnly: true,
    };

    const wrapper = shallow(<UIInputTextReadOnly id="foo" uiField={roNickName} />);

    expect(wrapper.find('UIFormLabel')).toHaveLength(1);
    expect(wrapper.contains('Kerman')).toEqual(true);
    expect(wrapper.find('InputText')).toHaveLength(0);
    expect(wrapper.find('UIInputTextReadOnly')).toHaveLength(0);
  });

  it('Read Only Password Field', () => {
    const roNickName: UiStringField = {
      ...nickName,
      value: 'Kerman',
      readOnly: true,
    };

    const wrapper = shallow(<UIInputTextReadOnly id="foo" type="password" uiField={roNickName} />);

    expect(wrapper.find('UIFormLabel')).toHaveLength(1);
    expect(wrapper.contains('Kerman')).toEqual(false);
    expect(wrapper.contains('********')).toEqual(true);
    expect(wrapper.find('InputText')).toHaveLength(0);
    expect(wrapper.find('UIInputTextReadOnly')).toHaveLength(0);
  });

  it('Automatic Read Only rendering form UI Field', () => {
    const roNickName: UiStringField = {
      ...nickName,
      value: 'Kerman',
      readOnly: true,
    };

    const wrapper = shallow(<UIInputText id="foo" uiField={roNickName} value="Jimbo" />);

    expect(wrapper.find('InputText')).toHaveLength(0);
    expect(wrapper.find('UIInputTextReadOnly')).toHaveLength(1);
  });

  it('Hidden UI Field', () => {
    const hiddenField: UiStringField = {
      ...nickName,
      visible: false,
    };

    const wrapper = mountWithTheme(<UIInputText id="_Hidden" uiField={hiddenField} />);

    expect(wrapper.html()).toEqual('');
  });
});
