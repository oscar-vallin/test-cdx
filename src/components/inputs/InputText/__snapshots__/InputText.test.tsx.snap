// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches Snapshot 1`] = `
<Styled(StyledTextFieldBase)
  autofocus={true}
  disabled={false}
  errorMessage={null}
  id="InputText"
  onChange={[Function]}
  onKeyDown={[Function]}
  placeholder="This is a placeholder"
  type="text"
  value=""
/>
`;
